runtime/jslib.js

```diff
@@ -50,6 +51,67 @@ function caml_trampoline_return(f,args) {
   return {joo_tramp:f,joo_args:args};
 }
 
+//Provides:caml_stack_depth
+//If: effects
+var caml_stack_depth = 0;
+
+//Provides:caml_stack_check_depth
+//If: effects
+//Requires:caml_stack_depth
+function caml_stack_check_depth() {
+    return --caml_stack_depth > 0;
+}
+
+//Provides: caml_callback
+//If: !effects
+//Requires:caml_call_gen
+var caml_callback = caml_call_gen;
+
+//Provides: caml_callback
+//If: effects
+//Requires:caml_stack_depth, caml_call_gen, caml_exn_stack, caml_fiber_stack, caml_wrap_exception, caml_resume_stack, caml_fresh_oo_id, caml_named_value, caml_raise_with_arg, caml_string_of_jsbytes
+//Requires: caml_raise_constant
+function caml_callback(f,args) {
+  function uncaught_effect_handler(eff,k,ms) {
+    // Resumes the continuation k by raising exception Unhandled.
+    caml_resume_stack(k[1],ms);
+    var exn = caml_named_value("Effect.Unhandled");
+    if(exn) caml_raise_with_arg(exn, eff);
+    else {
+      exn = [248,caml_string_of_jsbytes("Effect.Unhandled"), caml_fresh_oo_id(0)];
+      caml_raise_constant(exn);
+    }
+  }
+  var saved_stack_depth = caml_stack_depth;
+  var saved_exn_stack = caml_exn_stack;
+  var saved_fiber_stack = caml_fiber_stack;
+  try {
+    caml_exn_stack = 0;
+    caml_fiber_stack =
+      {h:[0, 0, 0, uncaught_effect_handler], r:{k:0, x:0, e:0}};
+    var res = {joo_tramp: f,
+               joo_args: args.concat(function (x){return x;})};
+    do {
+      caml_stack_depth = 40;
+      try {
+        res = caml_call_gen(res.joo_tramp, res.joo_args);
```

Comment:
Isn't it too expensive to setup an exception handler every time we bounce on the trampoline ? We could have a while-loop inside the `try` for fast rebound

Reply:
I think setting up an exception handler is cheap now. For a long time, V8 was not able to generate native code for function containing exception handlers, but that was fixed when they switch from Crankshaft to TurboFan about five years ago.

Reply:
Isn't the boyer benchmark still showing significant performance impact when using exception ? Is it only throwing exception that is expensive ?

Reply:
I think so, yes